auth/auth0_manager_test.go

```diff
@@ -0,0 +1,144 @@
+package auth
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAuth0ManagerAuthenticate_ValidCreds(t *testing.T) {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, "POST")
+		assert.Equal(t, r.URL.Path, "/oauth/ro")
+
+		var req oauthReq
+		Unmarshal(t, r, &req)
+
+		assert.Equal(t, req.Username, "valid username")
+		assert.Equal(t, req.Password, "valid password")
+
+		MarshalAndWrite(t, w, nil, 200)
+	}
+
+	auth0Manager, server := newAuth0ManagerAndServer(handler)
+	defer server.Close()
+
+	valid, err := auth0Manager.Authenticate("valid username", "valid password")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, valid, true)
+}
+
+func TestAuth0ManagerAuthenticate_InvalidCreds(t *testing.T) {
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, "POST")
+		assert.Equal(t, r.URL.Path, "/oauth/ro")
+
+		var req oauthReq
+		Unmarshal(t, r, &req)
+
+		assert.Equal(t, req.Username, "valid username")
+		assert.Equal(t, req.Password, "invalid password")
+
+		MarshalAndWrite(t, w, nil, 401)
+	}
+
+	auth0Manager, server := newAuth0ManagerAndServer(handler)
+	defer server.Close()
+
+	valid, err := auth0Manager.Authenticate("valid username", "invalid password")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, valid, false)
+}
+
+func TestAuth0ManagerAuthenticate_GoodCredsTurnedBad(t *testing.T) {
+	count := 0
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, "POST")
+		assert.Equal(t, r.URL.Path, "/oauth/ro")
+
+		var req oauthReq
+		Unmarshal(t, r, &req)
+
+		if count == 0 {
+			assert.Equal(t, req.Username, "valid username")
+			assert.Equal(t, req.Password, "valid password")
+
+			MarshalAndWrite(t, w, nil, 200)
+		} else {
+			assert.Equal(t, req.Username, "valid username")
+			assert.Equal(t, req.Password, "invalid password")
+
+			MarshalAndWrite(t, w, nil, 401)
+		}
+
+		count++
+	}
+
+	auth0Manager, server := newAuth0ManagerAndServer(handler)
+	defer server.Close()
+
+	valid, err := auth0Manager.Authenticate("valid username", "valid password")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, valid, true)
+
+	valid, err = auth0Manager.Authenticate("valid username", "invalid password")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, valid, false)
+}
+
+func TestAuth0ManagerAuthenticate_BadCredsTurnedGood(t *testing.T) {
```

Comment:
Again, I don't think this is testing the behavior that you want. 

I would recommend testing the inverse of the scenario above:
* Call `auth0Manager.Authenticate()`, and have the handler return a `401`
* Call `auth0Manager.Authenticate()` again (with the same creds!), and this time have the handler return a `200`. 
If the handler wasn't called twice, the test should fail. 

To make sure this test runs fast, I would recommend adding a package level variable, `var timeMultiplier = 1`. 
Update `time.Sleep(cachedStatus.penalty)` to `time.Sleep(cachedStatus.penalty * timeMultiplier)`, and during testing make sure to set `timeMultiplier = 0`.